c10/core/impl/LocalDispatchKeySet.cpp

```diff
@@ -47,50 +47,33 @@ void _force_tls_local_dispatch_key_set(LocalDispatchKeySet key_set) {
 }
 
 // An RAII guard could snapshot and restore the entire state (entire DispatchKeySet) as
-// opposed to only snapshotting and restoring the state of its assigned DispatchKey.
+// opposed to only snapshotting and restoring the state of its assigned DispatchKeySet.
 // I'm not sure which is better.  If only the RAII API is used, the two choices are
 // not distinguishable.
 //
 // However, if the guard chooses to snapshot and restore the entire DispatchKeySet,
 // the interaction with the non-RAII API changes.  Consider this sequence of events:
-// - An RAII guard is declared for a particular DispatchKey, but snapshots the entire
+// - An RAII guard is declared for a particular DispatchKeySet, but snapshots the entire
 //   current DispatchKeySet.
-// - A call to the non-RAII API changes the state for a different DispatchKey.
+// - A call to the non-RAII API changes the state for DispatchKeys outside the assigned
+//   set.
 // - The RAII guard goes out of scope, restoring the entire DispatchKeySet it snapshotted
 //   (which restores the state for its own assigned DispatchKey and wipes out the state
-//   for the other DispatchKey set by the non-RAII API).
+//   for the other DispatchKeys set by the non-RAII API).
 
 // RAII API
 
-IncludeDispatchKeyGuard::IncludeDispatchKeyGuard(DispatchKey x)
+IncludeDispatchKeySetGuard::IncludeDispatchKeySetGuard(DispatchKeySet include)
   : tls_(&raw_local_dispatch_key_set)
-  , id_(x)
-  // NB: prev_state_ == true on Undefined makes the guard no-op
-  , prev_state_(x == DispatchKey::Undefined ? true : tls_->included().has(x)) {
-  if (!prev_state_) {
-    tls_->set_included(tls_->included().add(x));
+  , include_(include - tls_->included()) {
```

Comment:
Note that this is subtly semantically different than the old version, which can be observed if there is a manual change to the dispatch key set that is not RAII guarded.

Previously:

```
// suppose K1 is included
{
  IncludeDispatchKeyGuard g(K1);
  // remove K1 from included
}
// K1 is back in included in the end!
```

After your change, K1 will still be missing in the end

Reply:
Are you sure that's how it used to work? Per the big comment above, I think the old logic also ignored changes made outside the include/exclude API (i.e., didn't restore a snapshot).

Just to make sure I'm not smoking crack...
```
        old                                    new
ctor    prev_state = tls_->included has key    keyset -= tls_->included
        if !prev_state tls_->included += key   tls_->included |= keyset
dtor    if !prev_state tls_->included -= key   tls_->included -= keyset
```
If this is the right reading of the old code, I think it ignored behind-the-back changes too.

Reply:
Oh, I guess you're right! My bad.